[JSC] Move MicrotaskQueue driving AsyncGenerator code to C++ and remove "WithoutPromise" variants

Constellation opened the pull request:
#### c88ec0937ab27555a59e986679429f52f1c90b30
<pre>
[JSC] Move MicrotaskQueue driving AsyncGenerator code to C++ and remove &quot;WithoutPromise&quot; variants
<a href="https://bugs.webkit.org/show_bug.cgi?id=304772">https://bugs.webkit.org/show_bug.cgi?id=304772</a>
<a href="https://rdar.apple.com/167318733">rdar://167318733</a>

Reviewed by Sosuke Suzuki.

This patch adds MicrotaskQueue driving AsyncGenerator code in C++. But
importantly, we still keep JS version of that. The reason is that these
code can be called synchronously from `next()` call of AsyncGenerator,
and in this case, keeping them in JS is better for performance when it
is calling AsyncGenerator&apos;s body function. This also simplifies JS side
of AsyncGenerator code: we can inline `doAsyncGeneratorBodyCall` into
`asyncGeneratorResumeNext`.

As a result, we no longer need any &quot;WithoutPromise&quot; variants of
JSPromise related functions. We remove them since they can be replaced
with InternalMicrotask version instead.

* Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js:
(linkTimeConstant.asyncGeneratorResumeNext):
(linkTimeConstant.asyncGeneratorYieldAwaited): Deleted.
(linkTimeConstant.asyncGeneratorYieldOnRejected): Deleted.
(linkTimeConstant.asyncGeneratorYield):
(linkTimeConstant.doAsyncGeneratorBodyCallOnFulfilledNormal): Deleted.
(linkTimeConstant.doAsyncGeneratorBodyCallOnFulfilledReturn): Deleted.
(linkTimeConstant.doAsyncGeneratorBodyCall):
(linkTimeConstant.asyncGeneratorResumeNextOnFulfilled): Deleted.
(linkTimeConstant.asyncGeneratorResumeNextOnRejected): Deleted.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.cpp:
(JSC::BytecodeIntrinsicRegistry::BytecodeIntrinsicRegistry):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSAsyncGenerator.h:
* Source/JavaScriptCore/runtime/JSFunctionWithFields.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::iteratorProtocolFunction const):
(JSC::JSGlobalObject::asyncGeneratorYieldOnRejectedFunction const): Deleted.
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::asyncGeneratorQueueIsEmpty):
(JSC::asyncGeneratorQueueDequeue):
(JSC::isSuspendYieldState):
(JSC::asyncGeneratorReject):
(JSC::asyncGeneratorResolve):
(JSC::asyncGeneratorYield):
(JSC::doAsyncGeneratorBodyCall):
(JSC::asyncGeneratorResumeNext):
(JSC::asyncGeneratorYieldAwaited):
(JSC::asyncGeneratorBodyCallNormal):
(JSC::asyncGeneratorBodyCallReturn):
(JSC::asyncGeneratorResumeNextReturn):
(JSC::runInternalMicrotask):
(JSC::promiseResolveThenableJobWithoutPromiseFastSlow): Deleted.
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSPromise::performPromiseThen):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSPromise::triggerPromiseReactions):
(JSC::JSPromise::resolveWithoutPromiseForAsyncAwait): Deleted.
(JSC::JSPromise::resolveWithoutPromise): Deleted.
(JSC::JSPromise::rejectWithoutPromise): Deleted.
(JSC::JSPromise::fulfillWithoutPromise): Deleted.
* Source/JavaScriptCore/runtime/JSPromise.h:
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::promiseRaceSlow):
* Source/JavaScriptCore/runtime/Microtask.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::promiseResolvingFunctionResolveWithInternalMicrotaskExecutableSlow):
(JSC::VM::promiseResolvingFunctionRejectWithInternalMicrotaskExecutableSlow):
(JSC::VM::visitAggregateImpl):
(JSC::VM::promiseResolvingFunctionResolveWithoutPromiseExecutableSlow): Deleted.
(JSC::VM::promiseResolvingFunctionRejectWithoutPromiseExecutableSlow): Deleted.
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::promiseResolvingFunctionResolveWithInternalMicrotaskExecutable):
(JSC::VM::promiseResolvingFunctionRejectWithInternalMicrotaskExecutable):
(JSC::VM::promiseResolvingFunctionResolveWithoutPromiseExecutable): Deleted.
(JSC::VM::promiseResolvingFunctionRejectWithoutPromiseExecutable): Deleted.

Canonical link: <a href="https://commits.webkit.org/304998@main">https://commits.webkit.org/304998@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57f75b2d008f67e29474eb5c5eb04a8ca3078099

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137143 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9503 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48430 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144889 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90116 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7d0812e9-43f2-4e4c-afda-a892acdf5c21) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10206 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9630 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104875 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76498 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3ca6f94e-99b8-4db2-8487-a25f7bc398b6) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140088 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7510 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/122898 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85713 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8d406da3-2be0-4859-a4dc-a75942c73982) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7147 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4858 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5479 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/129104 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116498 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41065 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147645 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/135632 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9186 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41628 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113233 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9204 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7726 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113563 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7068 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119158 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63563 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21131 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9234 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37212 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/168411 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8958 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72800 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43957 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9175 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9027 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js:
```diff
@@ -120,60 +120,54 @@ function asyncGeneratorResolve(generator, value, done)
-function doAsyncGeneratorBodyCall(generator, resumeValue, resumeMode)
-{
-    "use strict";
+            return @asyncGeneratorReject(generator, resumeValue);;
```
extra ;
```suggestion
            return @asyncGeneratorReject(generator, resumeValue);
```

Constellation replied:
Fixed.

sosukesuzuki reviewed (approved):
r=me, if EWS and test262 are happy

Constellation committed:
[JSC] Move MicrotaskQueue driving AsyncGenerator code to C++ and remove "WithoutPromise" variants
https://bugs.webkit.org/show_bug.cgi?id=304772
rdar://167318733

Reviewed by Sosuke Suzuki.

This patch adds MicrotaskQueue driving AsyncGenerator code in C++. But
importantly, we still keep JS version of that. The reason is that these
code can be called synchronously from `next()` call of AsyncGenerator,
and in this case, keeping them in JS is better for performance when it
is calling AsyncGenerator's body function. This also simplifies JS side
of AsyncGenerator code: we can inline `doAsyncGeneratorBodyCall` into
`asyncGeneratorResumeNext`.

As a result, we no longer need any "WithoutPromise" variants of
JSPromise related functions. We remove them since they can be replaced
with InternalMicrotask version instead.

* Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js:
(linkTimeConstant.asyncGeneratorResumeNext):
(linkTimeConstant.asyncGeneratorYieldAwaited): Deleted.
(linkTimeConstant.asyncGeneratorYieldOnRejected): Deleted.
(linkTimeConstant.asyncGeneratorYield):
(linkTimeConstant.doAsyncGeneratorBodyCallOnFulfilledNormal): Deleted.
(linkTimeConstant.doAsyncGeneratorBodyCallOnFulfilledReturn): Deleted.
(linkTimeConstant.doAsyncGeneratorBodyCall):
(linkTimeConstant.asyncGeneratorResumeNextOnFulfilled): Deleted.
(linkTimeConstant.asyncGeneratorResumeNextOnRejected): Deleted.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.cpp:
(JSC::BytecodeIntrinsicRegistry::BytecodeIntrinsicRegistry):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSAsyncGenerator.h:
* Source/JavaScriptCore/runtime/JSFunctionWithFields.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::iteratorProtocolFunction const):
(JSC::JSGlobalObject::asyncGeneratorYieldOnRejectedFunction const): Deleted.
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::asyncGeneratorQueueIsEmpty):
(JSC::asyncGeneratorQueueDequeue):
(JSC::isSuspendYieldState):
(JSC::asyncGeneratorReject):
(JSC::asyncGeneratorResolve):
(JSC::asyncGeneratorYield):
(JSC::doAsyncGeneratorBodyCall):
(JSC::asyncGeneratorResumeNext):
(JSC::asyncGeneratorYieldAwaited):
(JSC::asyncGeneratorBodyCallNormal):
(JSC::asyncGeneratorBodyCallReturn):
(JSC::asyncGeneratorResumeNextReturn):
(JSC::runInternalMicrotask):
(JSC::promiseResolveThenableJobWithoutPromiseFastSlow): Deleted.
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSPromise::performPromiseThen):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSPromise::triggerPromiseReactions):
(JSC::JSPromise::resolveWithoutPromiseForAsyncAwait): Deleted.
(JSC::JSPromise::resolveWithoutPromise): Deleted.
(JSC::JSPromise::rejectWithoutPromise): Deleted.
(JSC::JSPromise::fulfillWithoutPromise): Deleted.
* Source/JavaScriptCore/runtime/JSPromise.h:
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::promiseRaceSlow):
* Source/JavaScriptCore/runtime/Microtask.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::promiseResolvingFunctionResolveWithInternalMicrotaskExecutableSlow):
(JSC::VM::promiseResolvingFunctionRejectWithInternalMicrotaskExecutableSlow):
(JSC::VM::visitAggregateImpl):
(JSC::VM::promiseResolvingFunctionResolveWithoutPromiseExecutableSlow): Deleted.
(JSC::VM::promiseResolvingFunctionRejectWithoutPromiseExecutableSlow): Deleted.
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::promiseResolvingFunctionResolveWithInternalMicrotaskExecutable):
(JSC::VM::promiseResolvingFunctionRejectWithInternalMicrotaskExecutable):
(JSC::VM::promiseResolvingFunctionResolveWithoutPromiseExecutable): Deleted.
(JSC::VM::promiseResolvingFunctionRejectWithoutPromiseExecutable): Deleted.

Canonical link: https://commits.webkit.org/304998@main
